<a href="https://colab.research.google.com/github/maryandy2002/maryandy2002/blob/main/MetododeSecante.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Método de la Secante
El método de la secante es un método numérico utilizado para encontrar aproximaciones a las raíces de una función $f(x)$. Se basa en la idea de reemplazar la derivada en el método de Newton por una aproximación de la pendiente usando dos puntos consecutivos de la función.

Se realizara una aproximacion a la raiz de la funcion $f(x)=cos(x)-x$ utilizando el método de la secante con un $p_0=\frac{1}{2}$ y $p_1=\frac{\pi}{4}$

En este bloque importamos las librerías que se van a utilizar,
 para crear una tabla al final para visualizar los resultados y
 para codificar la función que usaremos para aproximar la raíz

In [32]:
from tabulate import tabulate as tb
import math as mt

Ahora se define la funcion $funcion$ que nos ayudara a calcular $f(x)$ fácilmente en cada una de nuestras iteraciones

In [37]:
def funcion(x):
  return mt.cos(x)-x

Aqui se definen nuestras listas
$n$ para ir contando la cantidad de iteraciones,
$e$ para ir calculando el error aproximado
$p_0$ y $p_1$ donde

$p_{i+1}=p_i-\frac{f(p_i)(p_i-p_{i-1})}{f(p_i)-f(p_{i-1})}$ posteriormente se comienza con las iteraciones del metodo y los resultados se muestran en una tabla al final

In [45]:
def Secante(p,pp,tol):
  n=[0]
  e=[0]
  p0=[p]
  p1=[pp]
  while (True):
    p1.append(p1[-1]-(funcion(p1[-1])*(p1[-1]-p0[-1])/(funcion(p1[-1])-funcion(p0[-1]))))
    p0.append(p1[-2])
    e.append(abs(p1[-1]-p1[-2]))
    if (e[-1]<tol):
     break
    n.append(n[-1]+1)
  table=[n,p0,p1,e]
  print(tb((zip(*table)),headers=["n","pi-1","pi","e"]))

In [46]:
Secante(0.5,mt.pi/4,1e-5)

  n      pi-1        pi            e
---  --------  --------  -----------
  0  0.5       0.785398  0
  1  0.785398  0.736384  0.049014
  2  0.736384  0.739058  0.002674
  3  0.739058  0.739085  2.70101e-05


Así, podemos concluir que la raiz para la funcion $cos(x)-x$ es en $x=0.739085$ con un $\varepsilon=0.00002$ el cual coincide con los resultados obtenidos en el código para el método de Newton